sharding/syncer/service_test.go

```diff
@@ -0,0 +1,234 @@
+	}
+
+	body := []byte{1, 2, 3, 4, 5}
+	shardID := big.NewInt(0)
+	chunkRoot := types.DeriveSha(sharding.Chunks(body))
+	period := big.NewInt(0)
+	proposerAddress := common.BytesToAddress([]byte{})
+	signer := &mockSigner{}
+
+	header := sharding.NewCollationHeader(shardID, &chunkRoot, period, &proposerAddress, nil)
+	sig, err := signer.Sign(header.Hash())
+	if err != nil {
+		t.Fatalf("Could not sign header: %v", err)
+	}
+
+	// Adds the signature to the header before calculating the hash used for db lookups.
+	header.AddSig(sig)
+
+	// Stores the collation into the inmemory kv store shardChainDB.
+	collation := sharding.NewCollation(header, body, nil)
+
+	shard := sharding.NewShard(shardID, shardChainDB)
+
+	if err := shard.SaveCollation(collation); err != nil {
+		t.Fatalf("Could not store collation in shardChainDB: %v", err)
+	}
+
+	syncer, err := NewSyncer(params.DefaultConfig, &mainchain.SMCClient{}, server, shardChainDB, 0)
+	if err != nil {
+		t.Fatalf("Unable to setup syncer service: %v", err)
+	}
+
+	feed := server.Feed(messages.CollationBodyRequest{})
+
+	go syncer.handleCollationBodyRequests(&mockSigner{}, feed)
+
+	go func() {
+		for {
+			select {
+			case <-syncer.ctx.Done():
+				return
+			default:
+				msg := p2p.Message{
+					Peer: p2p.Peer{},
+					Data: messages.CollationBodyRequest{
+						ChunkRoot: &chunkRoot,
+						ShardID:   shardID,
+						Period:    period,
+						Proposer:  &proposerAddress,
+					},
+				}
+				feed.Send(msg)
+			}
+		}
+	}()
+
+	<-syncer.responseSent
+	h.VerifyLogMsg(fmt.Sprintf("Received p2p request of type: %T", p2p.Message{}))
+	h.VerifyLogMsg(fmt.Sprintf("Responding to p2p request with collation with headerHash: %v", header.Hash().Hex()))
+	syncer.cancel()
+	// The context should have been canceled.
+	if syncer.ctx.Err() == nil {
+		t.Fatal("Context was not canceled")
+	}
+}
+
+// TODO: Move this to the utils package along with the handleServiceErrors
+// function.
+func TestHandleServiceErrors(t *testing.T) {
+
+	h := internal.NewLogHandler(t)
+	log.Root().SetHandler(h)
+
+	shardChainDB := database.NewShardKV()
+	shardID := 0
+	server, err := p2p.NewServer()
+	if err != nil {
+		t.Fatalf("Unable to setup p2p server: %v", err)
+	}
+
+	syncer, err := NewSyncer(params.DefaultConfig, &mainchain.SMCClient{}, server, shardChainDB, shardID)
+	if err != nil {
+		t.Fatalf("Unable to setup syncer service: %v", err)
+	}
+
+	go syncer.handleServiceErrors()
+
+	expectedErr := "testing the error channel"
+	complete := make(chan int)
+
+	go func() {
+		for {
+			select {
+			case <-syncer.ctx.Done():
+				return
+			default:
+				syncer.errChan <- errors.New(expectedErr)
+				complete <- 1
+			}
+		}
+	}()
+
+	<-complete
+	syncer.cancel()
+
+	// The context should have been canceled.
+	if syncer.ctx.Err() == nil {
+		t.Fatal("Context was not canceled")
+	}
+	// Right now we need to wait a little bit before the log is called.
+	// TODO: better way to do this? I know this is bad practice.
```

Comment:
How about extending the log helper to return a channel that blocks until a log message occurs?
See this: https://play.golang.org/p/cDmUegusk37

Reply:
Let's do this in a separate Issue/PR after merge